crates/pallet-domains/src/staking.rs

```diff
@@ -0,0 +1,366 @@
+    /// Current epoch index for the domain.
+    pub current_epoch_index: EpochIndex,
+    /// Total active stake for the current epoch.
+    pub current_total_stake: Balance,
+    /// Total stake for the next epoch.
+    pub next_total_stake: Balance,
+    /// Current operators for this epoch
+    pub current_operators: Vec<OperatorId>,
+    /// Operators for the next epoch.
+    pub next_operators: Vec<OperatorId>,
+}
+
+#[derive(TypeInfo, Debug, Encode, Decode, Clone, PartialEq, Eq)]
+pub struct OperatorConfig<Balance> {
+    pub signing_key: ExecutorPublicKey,
+    pub minimum_nominator_stake: Balance,
+    pub nomination_tax: Percent,
+}
+
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    MaximumOperatorId,
+    DomainNotInitialized,
+    InsufficientBalance,
+    BalanceFreeze,
+    MinimumOperatorStake,
+    UnknownOperator,
+    MinimumNominatorStake,
+    BalanceOverflow,
+}
+
+pub(crate) fn do_register_operator<T: Config>(
+    operator_owner: T::AccountId,
+    domain_id: DomainId,
+    amount: BalanceOf<T>,
+    config: OperatorConfig<BalanceOf<T>>,
+) -> Result<OperatorId, Error> {
+    DomainStakingSummary::<T>::try_mutate(domain_id, |maybe_domain_stake_summary| {
+        let operator_id = NextOperatorId::<T>::get();
+        let next_operator_id = operator_id.checked_add(1).ok_or(Error::MaximumOperatorId)?;
+        NextOperatorId::<T>::set(next_operator_id);
+
+        OperatorIdOwner::<T>::insert(operator_id, operator_owner.clone());
+
+        // reserve stake balance
+        ensure!(
+            amount >= T::MinOperatorStake::get(),
+            Error::MinimumOperatorStake
+        );
+
+        freeze_account_balance_to_operator::<T>(&operator_owner, operator_id, amount)?;
+
+        let domain_stake_summary = maybe_domain_stake_summary
+            .as_mut()
+            .ok_or(Error::DomainNotInitialized)?;
+
+        let OperatorConfig {
+            signing_key,
+            minimum_nominator_stake,
+            nomination_tax,
+        } = config;
+
+        let operator = OperatorPool {
+            signing_key,
+            current_domain_id: domain_id,
+            minimum_nominator_stake,
+            nomination_tax,
+            current_total_stake: Zero::zero(),
+            next_total_stake: Zero::zero(),
+            total_shares: Zero::zero(),
+            is_frozen: false,
+        };
+        OperatorPools::<T>::insert(operator_id, operator);
+        // update stake summary to include new operator for next epoch
+        domain_stake_summary.next_operators.push(operator_id);
+        // update pending transfers
+        PendingTransfers::<T>::append(
+            operator_id,
+            PendingTransfer {
+                nominator_id: operator_owner,
+                transfer: Transfer::Deposit(amount),
+            },
+        );
+
+        Ok(operator_id)
+    })
+}
+
+pub(crate) fn do_nominate_operator<T: Config>(
```

Comment:
Should we check if the nominator has nominated the same operator already?

Reply:
I dont see any benefit to check that explictly. We just need to ensure we freeze already staked amount + any new deposit. We currently do that. Did I miss anything that might be of concern to you ?

Reply:
I was thinking if the current implement forgets to use the `Nominators` storage item to create `Nominator` and check duplication here, this is resolved in the above comment.

I do have another concern that now the `nominate_operator` call can be used to either join a pool or nominate additional funds to a joined pool, for nominating additional funds there should not have the `MinimumNominatorStake` constraint as the nominator's total stake already satisfied that.

And it may also be better to have 2 dedicated calls instead if the condition of joining the pool and nominating additional funds will be changed and become different in the future (i.e. letting the pool creator decide whether accept a new nominator to join), non-blocker.

Reply:
Yes, that is a redundant check. Migration path is much simpler from current impl to next but ATM that is out of scope for initial version. We can revisit of these again in the further milestones